app/models/agents/data_output_agent.rb

```diff
@@ -175,6 +180,60 @@ def push_hubs
       interpolated['push_hubs'].presence || []
     end
 
+    DEFAULT_EVENTS_ORDER = {
+      'events_order' => nil,
+      'events_list_order' => [["{{_index_}}", "number", true]],
+    }
+
+    def events_order(key = SortableEvents::EVENTS_ORDER_KEY)
+      super || DEFAULT_EVENTS_ORDER[key]
+    end
+
+    def latest_events(reload = false)
+      events =
+        if (event_ids = memory[:event_ids]) &&
+           memory[:events_order] == events_order &&
+           memory[:events_to_show] >= events_to_show
+          received_events.where(id: event_ids).to_a
+        else
+          memory[:last_event_id] = nil
+          reload = true
+          []
+        end
+
+      if reload
+        memory[:events_order] = events_order
+        memory[:events_to_show] = events_to_show
+
+        new_events =
+          if last_event_id = memory[:last_event_id]
+            received_events.where(Event.arel_table[:id].gt(last_event_id)).
+              order(id: :asc).to_a
+          else
+            source_ids.flat_map { |source_id|
+              # dig twice as many events as the number of
+              # `events_to_show`
+              received_events.where(agent_id: source_id).
+                last(2 * events_to_show)
+            }.sort_by(&:id)
+          end
+
+        unless new_events.empty?
+          memory[:last_event_id] = new_events.last.id
+          events.concat(new_events)
```

Comment:
Yes, exactly.  This is the only place `memory[:last_event_id]` is set to a non-nil value, and it will be greater than or equal to any value in `memory[:event_ids]`.  `new_events` are selected by `id > memory[:last_event_id]` (see above) if `memory[:last_event_id]` is non-nil, so there will be no duplicates between `events` and `new_events`
If `memory[:last_event_id]` is nil, `memory[:event_ids]` is assured to be nil or empty and therefore `events` is empty, and there will be no duplicates.